File: src/IsbInterop/Accessory/Wrappers/ForEach.vb
﻿Namespace Accessory.Wrappers
  ''' <summary>
  ''' Обертка над IForEach.
  ''' </summary>
  Friend MustInherit Class ForEach(Of TI As IIsbComObjectWrapper)
    Inherits IsbComObjectWrapper
    Implements IForEach(Of TI)
    ''' <summary>
    ''' Количество элементов в списке.
    ''' </summary>
    Public ReadOnly Property Count() As Integer Implements IForEach(Of TI).Count
      Get
        Return CInt(Me.GetRcwProperty("Count"))
      End Get
    End Property

    ''' <summary>
    ''' Признак конца набора данных.
    ''' </summary>
    Public ReadOnly Property EOF() As Boolean Implements IForEach(Of TI).EOF
      Get
        Return CBool(Me.GetRcwProperty("EOF"))
      End Get
    End Property

    ''' <summary>
    ''' Получить объект набора данных.
    ''' </summary>
    ''' <returns>Реквизит.</returns>
    Public Function GetValue() As TI Implements IForEach(Of TI).GetValue
      Dim curentValueRcw = Me.GetRcwProperty("Value")
      Dim value = IsbObjectResolver.Resolve(Of TI)(curentValueRcw)

      Return value
    End Function

    ''' <summary>
    ''' Перейти к следующему объекту набора данных.
    ''' </summary>
    Public Sub [Next]() Implements IForEach(Of TI).[Next]
      Me.InvokeRcwInstanceMethod("Next")
    End Sub

    ''' <summary>
    ''' Устанавливить указатель текущего элемента списка на первый элемент списка.
    ''' </summary>
    Public Sub Reset() Implements IForEach(Of TI).Reset
      Me.InvokeRcwInstanceMethod("Reset")
    End Sub

    ''' <summary>
    ''' Конструктор.
    ''' </summary>
    ''' <param name="contentsObject">Объект IContents.</param>
    Protected Sub New(contentsObject As Object)
      MyBase.New(contentsObject)
    End Sub
  End Class
End Namespace

File: src/IsbInterop.CSharp/Accessory/Wrappers/ForEach.cs
namespace IsbInterop.Accessory.Wrappers
{
  /// <summary>
  /// Обертка над IForEach.
  /// </summary>
  internal abstract class ForEach<TI> : IsbComObjectWrapper, IForEach<TI> where TI : IIsbComObjectWrapper
  {
    /// <summary>
    /// Количество элементов в списке.
    /// </summary>
    public int Count => (int)GetRcwProperty("Count");

    /// <summary>
    /// Признак конца набора данных.
    /// </summary>
    public bool EOF => (bool)GetRcwProperty("EOF");

    /// <summary>
    /// Значение текущего элемента списка.
    /// </summary>
    public TI Value
    {
      get
      {
        var rcwObject = GetRcwProperty("Value");
        var value = IsbObjectResolver.Resolve<TI>(rcwObject, Scope);

        return value;
      }
    }

    /// <summary>
    /// Перейти к следующему объекту набора данных.
    /// </summary>
    public void Next()
    {
      this.InvokeRcwInstanceMethod("Next");
    }

    /// <summary>
    /// Устанавливить указатель текущего элемента списка на первый элемент списка.
    /// </summary>
    public void Reset()
    {
      this.InvokeRcwInstanceMethod("Reset");
    }

    /// <summary>
    /// Конструктор.
    /// </summary>
    /// <param name="contentsObject">Объект IContents.</param>
    /// <param name="scope">Область видимости.</param>
    protected ForEach(object contentsObject, IScope scope) : base(contentsObject, scope) { }
  }
}
